docs/3-cloud-computing/3.2.5-lambda.md

```diff
@@ -0,0 +1,125 @@
+    - Make sure to have the right permissions
+    - Add a trigger of type s3 bucket
+        - Specify an event type
+            - Hint: Think of what kind of files you will be uploading to the bucket
+
+    - Make a lambda function that grabs every item from a csv file
+
+    Here is some helpful documentation for [boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/dynamodb.html)
+
+``` python
+
+import json
+import boto3
+import csv
+
+      # lambda function handler here
+
+   record_list = []
+   
+   try:
+      # boto3 resources for s3 here
+
+      # boto3 resource for dynamodb here
+       
+      table = dynamodb.Table('** DYNAMO-DB-TABLE-NAME ***')
+      bucket = event['Records'][0]['s3']['bucket']['name']
+      key = event['Records'][0]['s3']['object']['key']
+      s3_object = s3_resource.Object(bucket,key)
+      
+      data = s3_object.get()['Body'].read().decode("utf-8").splitlines()
+      lines=csv.reader(data)
+      headers=next(lines)
+      
+      for user_data in lines:
+         print(user_data)
+
+         # Create table items here
+      
+   except Exception as e:
+      print(str(e))
+            
+   return {
+      'statusCode': 200,
+      'body': json.dumps('CSV success')
+   }
+
+```
+
+4. Once your python code is ready, download this csv file [zillow.csv](img3/zillow.csv)
+
+5. Upload the csv file to the bucket you created at the beginning of the exercise.
+
+6. After uploading the file, the lambda function should run. if you're having issues make sure to check the logs (cloud watch), and go over your code.
+
+## Exercise - Creating a Lambda function to spin up ec2 instances
+
+1. Create a Lambda function and verify it was created in the aws console.
+2. Next we want to create a DynamoDB database and table.  To do this go to the DynamoDB service page and click `"create table"`.  Create an appropriate name and make the partition key something that will denote different table entries such as "index" or "instance".
+3. Create an EC2 instance and tag it with Key: `test-<your_name>-lambda` Value: `True`  or something else that distinguishes your instance from any others.
+4. In your Lambda function create:
+ - An EC2 client
+ - A Dynamodb resource
+ - A Table connected to a specific table you have created in that
+   Dynamodb resource.
+
+``` python
+import json
+import boto3
+import csv
+
+def lambda_handler(event, context):
+   
+   try: 
+      region = 'us-east-1'
+      ec2 = boto3.client('ec2', region_name=region)
+      dynamodb = boto3.resource('dynamodb')
+      table = dynamodb.Table('<your_table>')
+      
+     
+   except Exception as e:
+      print(str(e))
+            
+   return {
+      'statusCode': 200,
+      'body': json.dumps('It got to the end of the function')
+    }
+```
+
+7. Collect and print out the data on the instance we created earlier to see the datas structure as well as some of the available information.
```

Comment:
This jumps from 4 to 7 in numbering, but also I'm not quite clear what this is asking for. How does the dynamo db setup correspond to the ec2 instance we created. Are we loading the data from ec2 into the table. Can that be clarified?

Reply:
If it's just for use in the next section, it might be worth clarifying that, or moving the creation of that table until then.